### ISSUE TITLE
Firebase Authentication's error code seems to be broken after upgrading to 5.4.0  

### ISSUE THREAD

**Issue by mono0926** [firebase/firebase-ios-sdk#2522]
### Describe your environment

  * Xcode version: Version 10.1 (10B61, latest)
  * Firebase SDK version: 5.18.0 (latest)
  * Firebase Component: Auth (5.4.0, latest)

### Describe the problem

My app's auth logic was broken after upgrading to Firebase Authentication 5.4.0, so I investigated that.
It seems that `FIRAuthErrorCodeInternalError(17999)` error started occurring for many requests.

#### Steps to reproduce:

This is one of the steps to reproduce the issue.

1. Login with anonymous Firebase user: success.
2. Tried to link the Firebase user to exiting linked Facebook account.

After that, `FIRAuthErrorCodeCredentialAlreadyInUse(17025)` should return, but after upgrading to Firebase Authentication 5.4.0, `FIRAuthErrorCodeInternalError(17999)`started returning.

```swift
import FirebaseAuth
import FBSDKLoginKit

let firUser = Auth.auth().currentUser!
FBSDKLoginManager().logIn(withReadPermissions: ["public_profile"], from: viewController) { (result, error) in
    let fbToken = result!.token!

    let credential = FacebookAuthProvider.credential(withAccessToken: fbToken.tokenString)

    firUser.linkAndRetrieveData(with: credential, completion: { (result, error) in
        // error code should be `FIRAuthErrorCodeCredentialAlreadyInUse(17025)`
        // error code is `FIRAuthErrorCodeInternalError(17999)` after upgrading to Firebase Authentication 5.4.0
    })
}
```

I suspect that #2405 is related? 🤔

**Comment by ryanwilson** [comment]
@mono0926 thanks for the reply, @renkelvin or @Yue-Wang-Google can you please take a look?

**Comment by renkelvin** [comment]
@mono0926 Thanks for reporting! It's verified that the issue was introduced in #2405 and we're working on the fix. Thanks!

**Comment by Yue-Wang-Google** [comment]
@mono0926 could you please try #2530 and see if it works for you? It works for me.

**Comment by Yue-Wang-Google** [comment]
@mono0926 I'll close this bug for now as it works for me plus we're freezing the code tonight. 
Feel free to comment below whether it resolves your problem or not.

**Comment by mono0926** [comment]
@Yue-Wang-Google 

Thanks, I checked that, but the issue seems to be not fixed unfortunately, so I'd like you to reopen this issue.

I checked it by running the code which pasted here( https://github.com/firebase/firebase-ios-sdk/issues/2522#issue-419835788 ), and add breakpoints to the code you modified.

## v5.4.0 + #2530

`shortErrorMessage` is `MISSING_ID_TOKEN`.

<img width="825" alt="Screen Shot 2019-03-13 at 14 39 06" src="https://user-images.githubusercontent.com/1255062/54256181-7add8d00-459e-11e9-8929-c576fb59a4ab.png">
<img width="679" alt="Screen Shot 2019-03-13 at 14 40 33" src="https://user-images.githubusercontent.com/1255062/54256182-7b762380-459e-11e9-9996-1a3dab496775.png">

## v5.3.1

`shortErrorMessage` should be `FEDERATED_USER_ID_ALREADY_LINKED`.

<img width="776" alt="Screen Shot 2019-03-13 at 14 44 05" src="https://user-images.githubusercontent.com/1255062/54256183-7b762380-459e-11e9-8bff-fede351c011a.png">

<img width="646" alt="Screen Shot 2019-03-13 at 14 54 07" src="https://user-images.githubusercontent.com/1255062/54256719-44087680-45a0-11e9-8e24-3e38c4ad093f.png">

---

You tried to fix `kFederatedUserIDAlreadyLinkedMessage` specific problem at #2530, but It might be more global problem as I said 🤔 

> My app's auth logic was broken after upgrading to Firebase Authentication 5.4.0, so I investigated that.
It seems that FIRAuthErrorCodeInternalError(17999) error started occurring for many requests.

**Comment by Yue-Wang-Google** [comment]
Thanks for reporting.  #2530 did fix the issue when the user is new. However for existing users it seems the new code did return an wrong error. I'm going to reopen the issue. For now you could comment out FIRVerifyAssertionRequest.m's L154 (which set `body[kReturnIDPCredentialKey]`  to `@YES`) or simply revert to the old version. I'll fix the bug tomorrow.

**Comment by mono0926** [comment]
@Yue-Wang-Google 

> comment out FIRVerifyAssertionRequest.m's L154 (which set body[kReturnIDPCredentialKey] to @YES)

This works fine, thanks!!

<img width="647" alt="image" src="https://user-images.githubusercontent.com/1255062/54259334-bda46280-45a8-11e9-973b-1bcc0f8f37ee.png">

**Comment by Yue-Wang-Google** [comment]
I believe I found the official fix: https://github.com/firebase/firebase-ios-sdk/pull/2541
Please remember to 
1 revert  the comment in FIRVerifyAssertionRequest.m's L154 (set it back to YES) if you commented it out. kReturnIDPCredentialKey needs to be true in order to make OAuth working. 
2 You need to apply #2541 
3 You also need to apply the previous fix #2530  --- both changes are necessary. 


Let me know if that works for you. 

Thank you very much for early stage testing and reporting the bug to us!

**Comment by mono0926** [comment]
@Yue-Wang-Google 

> I believe I found the official fix: #2541

Oh, it should be fixed.

But, unfortunately without `FIRVerifyAssertionRequest.m's L154` change, the problem isn't fixed.

> 1 revert the comment in FIRVerifyAssertionRequest.m's L154 (set it back to YES) if you commented it out. kReturnIDPCredentialKey needs to be true in order to make OAuth working.

<img width="390" alt="Screen Shot 2019-03-13 at 16 44 23" src="https://user-images.githubusercontent.com/1255062/54261625-9a30e600-45af-11e9-9815-2833a89b1f4a.png">

> 2 You need to apply #2541

<img width="662" alt="Screen Shot 2019-03-13 at 16 45 27" src="https://user-images.githubusercontent.com/1255062/54261672-b5035a80-45af-11e9-8c65-ff58b2c13c7c.png">

> 3 You also need to apply the previous fix #2530 --- both changes are necessary.

<img width="794" alt="Screen Shot 2019-03-13 at 16 44 00" src="https://user-images.githubusercontent.com/1255062/54261698-c482a380-45af-11e9-9cef-0413a7cf4c4a.png">

Result:

<img width="698" alt="image" src="https://user-images.githubusercontent.com/1255062/54261959-64d8c800-45b0-11e9-8fd1-9fea391fff2e.png">

In summary, the results is below:

- #2530 + #2541: doesn't work
- #2530 + #2541 + `FIRVerifyAssertionRequest.m's L154` fix: works well
  - without #2541, it works fine as commented here: 
https://github.com/firebase/firebase-ios-sdk/issues/2522#issuecomment-472303143

**Comment by mono0926** [comment]
>> I believe I found the official fix: #2541

> Oh, it should be fixed.

I suspect that it was correct because initial implementation(May 16, 2017) used `response.IDToken` 🤔 

https://github.com/firebase/firebase-ios-sdk/blob/98ba64449a632518bd2b86fe8d927f4a960d3ddc/Firebase/Auth/Source/FIRUser.m#L884

**Comment by Yue-Wang-Google** [comment]
Sorry. I was drunk. Please try #2542 with #2530. 
I'm very confident that this is the right fix.

**Comment by mono0926** [comment]
>  Please try #2542 with #2530. (you need both patches, and you need to revert the kReturnIDPCredentialKey comment)

It works fine! Thanks for the quick fix 🙏

**Comment by Yue-Wang-Google** [comment]
> It works fine! Thanks for the quick fix 🙏

Great! I'll wait for it to be approved by my colleague tomorrow, and after that v5.4.1 will be released.

Thanks for your bug report and testing my fixes. I will also keep my promise to buy you a beer, by the way:)

**Comment by sentex** [comment]
I have the same problem. Get `FIRAuthErrorCodeInternalError(17999)` after upgrading to Firebase Authentication 5.4.0

**Comment by Yue-Wang-Google** [comment]
@sentex it's fixed in 5.4.1. See the above PRs

**Comment by sentex** [comment]
<img width="627" alt="Снимок экрана 2019-03-19 в 23 13 39" src="https://user-images.githubusercontent.com/3669767/54642237-cb705100-4a9c-11e9-9e0c-cb1871252319.png">

**Comment by sentex** [comment]
<img width="914" alt="Снимок экрана 2019-03-19 в 23 15 44" src="https://user-images.githubusercontent.com/3669767/54642310-f064c400-4a9c-11e9-8e84-ccfc541867e9.png">

**Comment by sentex** [comment]
how can I install this version ?

**Comment by Yue-Wang-Google** [comment]
@sentex you can wait a day or two and see if it is upgraded on Cocoapods. Or, you can simply patch in  #2542 and #2530.

**Comment by paulb777** [comment]
Or see instructions at https://github.com/firebase/firebase-ios-sdk#accessing-firebase-source-snapshots or wait an hour or so. We're publishing to[ CocoaPods](https://github.com/CocoaPods/Specs/commits/master) now.
